drafts/2019-10-01-rewriting-irewritable.md

```diff
@@ -0,0 +1,448 @@
+
+### Wart 1: `Children<T>`
+
+`Children<T>` is a custom struct, not a standard collection type:
+
+```csharp
+struct Children<T>
+{
+    public NumberOfChildren NumberOfChildren { get; }
+    public T First { get; }
+    public T Second { get; }
+    public ImmutableList<T> Many { get; }
+}
+enum NumberOfChildren
+{
+    None, One, Two, Many
+}
+```
+
+I wanted to avoid boxing when the object has a small number of children (which is fairly common in practice). So `GetChildren` returns a `Children<T>`, passing up to two children on the stack and the rest in an `ImmutableList`. The `NumberOfChildren` property tells the library how many of the struct's fields are filled in.
+
+This custom collection type is an extra hurdle to understanding `IRewritable`'s API. It also makes certain parts of Sawmill's implementation more complex --- many internal methods have to `switch` on the `NumberOfChildren` they're working with and do the same work in four three different ways. It's also relatively large for a `struct` (around four words) so there is a marginal performance cost associated with copying it around.
+
+### Wart 2: Collections
+
+In fact, having `GetChildren` return a collection at all is problematic, because it forces implementations to allocate memory. If your object has three or more children (and they're not already stored in an `ImmutableList`) then you have to create a new `ImmutableList` whenever `GetChildren` is called:
+
+```csharp
+class ThreeChildren : IRewritable<ThreeChildren>
+{
+    private readonly ThreeChildren _child1;
+    private readonly ThreeChildren _child2;
+    private readonly ThreeChildren _child3;
+
+    public Children<ThreeChildren> GetChildren()
+        => new Children<ThreeChildren>(
+            Children.Many,
+            ImmutableList.Create(_child1, _child2, _child3)
+        );
+    public ThreeChildren SetChildren(Children<ThreeChildren> newChildren)
+        => new ThreeChildren(
+            newChildren.Many[0],
+            newChildren.Many[1],
+            newChildren.Many[2]
+        )
+}
+```
+
+When Sawmill traverses a tree it typically calls `GetChildren()` for every node in the tree. That's a lot of throwaway `ImmutableList`s!
+
+One way to avoid creating all this garbage might be to redesign `IRewritable` to be buffer-oriented.
+
+```csharp
+interface IRewritable<T> where T : IRewritable<T>
+{
+    int CountChildren();
+    void GetChildren(T[] buffer);
+    T SetChildren(T[] newChildren);
+}
+```
+
+With this design, Sawmill passes an array into `GetChildren` (after calling `CountChildren` to find out how big the array needs to be) and asks the object to copy its children into the array. So implementations of `IRewritable` now no longer have to allocate memory for their return value. But this API is less safe --- if an `IRewritable` implementation stores a reference to the buffer then it could get unexpectedly mutated.
```

Comment:
"So ... now" reads weirdly.  Just " Omplementations of `IRewritable`  no longer have to allocate memory for their return value." maybe?